### ISSUE TITLE
Trusted Domain list is case-sensitive

### ISSUE THREAD

**Issue by UniversalSuperBox** [nextcloud/server#17377]
### Steps to reproduce
1. Set your desired trusted domain in `config/config.php` in all lowercase. For example, `array (0=>nc.example.com,)`
2. Use cURL to retrieve your login page from your trusted domain. For example, `curl -I https://nc.example.com/login`. Observe that you receive an HTTP 200 (OK) response.
3. Use cURL to retrieve your login page again, but capitalize part of your trusted domain. For example, `curl -I https://NC.example.com/login`.

### Expected behaviour

The request succeeds. `NC.example.com` and `nc.example.com` will resolve to the same host, DNS is not case sensitive as far as I can see.

### Actual behaviour

Nextcloud returns "400 Bad Request" with the HTML indicating that a trusted domain violation was detected.

As a sanity check, if you add `NC.example.com` to your trusted domains, your request will succeed.

### Server configuration

**Operating system**: Ubuntu 18.04.3

**Web server:** nginx/1.14.0 (Ubuntu)

**Database:** MariaDB `10.1.41-MariaDB-0ubuntu0.18.04.1 Ubuntu 18.04`

**PHP version:** `PHP 7.2.19-0ubuntu0.18.04.2`

**Nextcloud version:** 16.0.5

**Updated from an older Nextcloud/ownCloud or fresh install:** Server has been upgraded since 14.0

**Where did you install Nextcloud from:** ZIP downloaded from nextcloud.org

**Signing status:**
<details>
<summary>Signing status</summary>

```
No errors have been found.
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
  - accessibility: 1.2.0
  - activity: 2.9.1
  - bruteforcesettings: 1.4.0
  - calendar: 1.7.1
  - cloud_federation_api: 0.2.0
  - comments: 1.6.0
  - contacts: 3.1.4
  - dav: 1.9.2
  - deck: 0.6.6
  - federatedfilesharing: 1.6.0
  - federation: 1.6.0
  - files: 1.11.0
  - files_markdown: 2.1.0
  - files_pdfviewer: 1.5.0
  - files_rightclick: 0.15.1
  - files_sharing: 1.8.0
  - files_texteditor: 2.8.0
  - files_trashbin: 1.6.0
  - files_versions: 1.9.0
  - files_videoplayer: 1.5.0
  - firstrunwizard: 2.5.0
  - gallery: 18.3.0
  - logreader: 2.1.0
  - lookup_server_connector: 1.4.0
  - news: 14.0.0
  - notes: 3.0.3
  - notifications: 2.4.1
  - oauth2: 1.4.2
  - password_policy: 1.6.0
  - polls: 0.10.2
  - privacy: 1.0.0
  - provisioning_api: 1.6.0
  - recommendations: 0.4.0
  - serverinfo: 1.6.0
  - sharebymail: 1.6.0
  - support: 1.0.0
  - survey_client: 1.4.0
  - systemtags: 1.6.0
  - theming: 1.7.0
  - twofactor_backupcodes: 1.5.0
  - twofactor_totp: 3.0.1
  - twofactor_u2f: 3.0.1
  - updatenotification: 1.6.0
  - viewer: 1.1.0
  - workflowengine: 1.6.0
Disabled:
  - admin_audit
  - encryption
  - files_external
  - nextcloud_announcements
  - user_ldap
```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "nc.example.com", // Edited for safety
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "overwrite.cli.url": "https:\/\/nc.example.com", // Edited for safety
        "dbtype": "mysql",
        "version": "16.0.5.1",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "maintenance": false,
        "theme": "",
        "loglevel": 2,
        "mail_smtpmode": "smtp",
        "mail_smtpauthtype": "LOGIN",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpsecure": "ssl",
        "mail_smtpport": "465",
        "mail_smtpauth": 1,
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "mysql.utf8mb4": true,
        "updater.secret": "***REMOVED SENSITIVE VALUE***"
    }
}
```
</details>

**Are you using external storage, if yes which one:** No

**Are you using encryption:** No

**Are you using an external user-backend, if yes which one:** No

### Client configuration
**Browser:** `curl 7.64.0`

**Operating system:** Ubuntu MATE 19.04

#### Web server error log
<details>
<summary>Web server error log</summary>

```
- log is silent -
```
</details>

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>

```
- log is silent -
```
</details>

#### Browser log
<details>
<summary>Browser log</summary>

```
HTTP/2 400
```
</details>

I note that it's possible to send all URLs to lowercase in the web server, mostly by using Perl. That's not ideal, but it is a solution.

I tried to make sure this wasn't something reported on or tested... and it appears not. [TrustedDomainHelperTest.php](https://github.com/nextcloud/server/blob/2e36069e24406455ad3f3998aa25e2a949d1402a/tests/lib/Security/TrustedDomainHelperTest.php) does not have any uppercase letters.

Most of the time you won't notice this. Browsers (Firefox at least) coerce domain names to lowercase when they're typed in while leaving the request string untouched. XMLHttpRequest also coerces domain names to lowercase. I found this while debugging an older version of Evolution Data Server, pulling my hair out and yelling "EVERYTHING IS CORRECT, WHY IS THE REQUEST FAILING!?"

**Comment by johkoenig** [comment]
That should do the trick...

**Comment by johkoenig** [comment]
solved by #17431 
@gary-kim: Can this be closed or do we wait for the 18 release?
